http-client-clj/src/main/clojure/xtdb/client.clj

```diff
@@ -61,19 +60,32 @@
       (throw (NoSuchElementException.)))
     (let [el (.next-el this)]
       (set! (.next-el this) nil)
-      el))
+      (if (instance? xtdb.RuntimeException el)
+        (throw el)
+        el)))
 
   (close [_]
     (.close in)))
 
-(defmethod hato.middleware/coerce-response-body ::transit+json->resultset [_req {:keys [^InputStream body] :as resp}]
-  (try
-    (let [rdr (transit/reader body :json {:handlers xt.transit/tj-read-handlers})]
-      (-> resp
-          (assoc :body (TransitResultSet. body rdr nil))))
-    (catch Exception e
-      (.close body)
-      (throw e))))
+(defmethod hato.middleware/coerce-response-body ::transit+json->result-or-error [_req {:keys [^InputStream body status] :as resp}]
+  (letfn [(parse-body [rdr->body]
+            (try
+              (let [rdr (transit/reader body :json {:handlers xt.transit/tj-read-handlers})]
+                (-> resp (assoc :body (rdr->body rdr))))
+              (catch Exception e
+                (.close body)
+                (throw e))))]
+
+    (cond
+      (= 200 status)
+      (parse-body (fn [rdr] (TransitResultSet. body rdr nil)))
+
+      ;; This should be an error we know how to decode
+      (<= 400 status 499)
+      (parse-body transit/read)
+
+      ;; Error will be thrown by error handler as is
+      :else nil)))
```

Comment:
I don't know how to best handle other cases here. For the 5XX range should I also expect some `tranist+json`?